### PR TITLE
Implements extended if grammar

### PR DESCRIPTION
**Pull request by mrange**
Origin: https://fslang.uservoice.com/forums/245727-f-language/suggestions/6079342-allow-extended-if-grammar

This commit adds the possibility to write condtional compilation expressions like this:
 #if SILVERLIGHT || NETFX
 #endif

or a bit more advanced:
 #if (SILVERLIGHT || NETFX) && COMPILED || !DEBUG
 #endif

This commit doesn't add #elif which is an orthogonal problem.

### CODE REVIEW

**Comment by mrange** [comment]
Migrated from CodePlex (mainly because I wanted a clean-slate as the history of this PR is quite long)

**Comment by mrange** [comment]
After pouring through the F# parser on how to do error handling I have incorporated similar techniques now. The result should be a more stable preprocessor parser and with support for better error messages to the user.

I ran the test suites (including VS integration)

I saw no errors that were due to extended if grammar.

There are two "small" issues remaining:

Inside visual studio the squiggle for unrecognized characters are rendered in the wrong place:
#if COMPILED # COMPILED // Squiggle rendered here -->_
#endif

The error message from the compiler looks correct though but I think its due to that I consume all characters. If I go back to a previous pattern and simpler pattern it's probably fixed.

The unit test for: 
#if (COMPILED ||)
#endif

This generates a parse error (and indication the preprocessor parser needs to be extended to handle this case). I have code that are supposed to handle it but it doesn't pick it up properly yet.

Conclusion:
1. I added Positive test cases on the fsc level which test various scenarios
2. I added a new unit test container for internal compiler tests that test positive and negative scenarios
3. The visual studio unit tests are extended with some new test cases
4. I have tested the behavior om the VS addin in VS2013 and I think it behaves reasonable (with the exception of unrecognized chars)
5. I have tested from the command line and think it behaves reasonable (expect for the parse error for certain incomplete expressions.

Therefore to my current knowledge I think this could be good to merge and schedule some minor bug fixes for later.

File: src/fsharp/pplex.fsl [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+| "||"          { PPParser.OP_OR                                                }
+| "("           { PPParser.LPAREN                                               }
+| ")"           { PPParser.RPAREN                                               }
+// --------------------------
+| whitespace    { tokenstream args lexbuf                                       }
+// --------------------------
+| comment       { PPParser.EOF                                                  }
+| mcomment      { fail args lexbuf (FSComp.SR.pplexExpectedSingleLineComment()) }
+| _             {
+                    let lex = lexeme lexbuf
```
Comment (mrange) [inline src/fsharp/pplex.fsl]: Is suspect consuming "rest" is what causes the squiggle to be rendered in the wrong place on this error. Should be ok with just consuming '_'

File: src/fsharp/FSharp.Compiler.Unittests/HashIfExpression.fs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                "ONE@"
+                "@ONE"
+                "$"
+                "ONE$"
+                "$ONE"
+                "ONE!"
+                "(ONE"
+                "ONE)"
+                // TODO: Investigate why this raises a parse failure
+                // "(ONE ||)"
```
Comment (mrange) [inline src/fsharp/FSharp.Compiler.Unittests/HashIfExpression.fs]: Should obviously be fixed but when executing from the command line while it looks a bit odd to the 'parse error' the compiler refuses to compile (as it should) as well as it gives an indication the expression is incomplete at the right place.

File: vsintegration/src/unittests/Tests.LanguageService.ErrorList.fs [hunk 3]
```diff
@@ -580,14 +580,6 @@ but here has type
         VerifyErrorListContainedExpectedStr("nonexistent",project)
 
     [<Test>]
-    member public this.``ErrorReporting.WrongPreprocessorIf``() =
```
Comment (mrange) [inline vsintegration/src/unittests/Tests.LanguageService.ErrorList.fs]: Dropped as this is now legal

File: vsintegration/src/unittests/Tests.LanguageService.Squiggles.fs [hunk 4]
```diff
@@ -855,16 +855,12 @@ type X() =
         this.TestSquiggle true [ "#if"; "#endif" ] "if" "#if directive should be immediately followed by an identifier"
 
     [<Test>]
-    member public this.``Squiggles.HashIfWrongExpr``() =
```
Comment (mrange) [inline vsintegration/src/unittests/Tests.LanguageService.Squiggles.fs]: Dropped as this is now legal

File: tests/RunTests.cmd [hunk 5]
```diff
@@ -178,6 +182,22 @@ echo nunit-console.exe /nologo /result=%XMLFILE% /output=%OUTPUTFILE% /err=%ERRO
 
 goto :EOF
 
+:COMPILERUNIT
```
Comment (mrange) [inline tests/RunTests.cmd]: This unit test should be automatically run by CI but I am not sure where this is configured. The test suite is rather small (for now) it's quick.

File: tests/fsharpqa/Source/Conformance/LexicalAnalysis/ConditionalCompilation/ExtendedIfGrammar.fs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    +   verify e12  0x001000
+    +   verify e13  0x002000
+    +   verify e14  0x004000
+    +   verify e15  0x008000
+    +   verify e16  0x010000
+    +   verify e17  0x020000
+    +   verify e18  0x040000
+    +   verify e19  0x080000
+    +   verify e20  0x100000
+    +   verify e21  0x200000
```
Comment (mrange) [inline tests/fsharpqa/Source/Conformance/LexicalAnalysis/ConditionalCompilation/ExtendedIfGrammar.fs]: Oops, should be 0x400000 and 0x800000 respectively. But there's no risk of false positives due to this but it makes finding the failure a little harder.

File: tests/fsharpqa/Source/Conformance/LexicalAnalysis/ConditionalCompilation/ExtendedIfGrammar.fs [hunk 7]
```diff
@@ -0,0 +1,182 @@
+// #Conformance #LexicalAnalysis
+
+let success = 1
```
Comment (mrange) [inline tests/fsharpqa/Source/Conformance/LexicalAnalysis/ConditionalCompilation/ExtendedIfGrammar.fs]: These tests might seem a bit unnecessary as the unit tests cover these and more, however it's good IMO to have tests on the same level as the user are operating (ie fsi/fsc)

Comment (mrange) [inline tests/fsharpqa/Source/Conformance/LexicalAnalysis/ConditionalCompilation/ExtendedIfGrammar.fs]: Could be reduced in scope though.

File: src/fsharp/FSharp.Compiler.Unittests/HashIfExpression.fs [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+open Microsoft.FSharp.Compiler.Lexer
+open Microsoft.FSharp.Compiler.Lexhelp
+open Microsoft.FSharp.Compiler.ErrorLogger
+open Microsoft.FSharp.Compiler.ErrorLogger
+open Microsoft.FSharp.Compiler.Ast
+
+[<TestFixture>]
+type HashIfExpression()     =
+
+    let preludes    = [|"#if "; "#elif "|]
```
Comment (mrange) [inline src/fsharp/FSharp.Compiler.Unittests/HashIfExpression.fs]: Hmm negative test cases should test missing/invalid prelude.

Comment (mrange) [inline src/fsharp/FSharp.Compiler.Unittests/HashIfExpression.fs]: Although that won't happen in the compiler currently because lex.fsl guarantee correct prelude.

**Comment by mexx** [comment]
Please consider to revert the changes in `lexhelp.fs` and `lexhelp.fsi` as they're only formatting and don't add any real value.

File: src/fsharp/FSharp.LanguageService.Compiler/pplex.fs [hunk 9]
```diff
@@ -0,0 +1,178 @@
+# 3 "..\pplex.fsl"
```
Comment (dsyme) [inline src/fsharp/FSharp.LanguageService.Compiler/pplex.fs]: This file should not be included in the commit

File: src/fsharp/FSharp.LanguageService.Compiler/pppars.fs [hunk 10]
```diff
@@ -0,0 +1,374 @@
+// Implementation file for parser generated by fsyacc
```
Comment (dsyme) [inline src/fsharp/FSharp.LanguageService.Compiler/pppars.fs]: This file should not be included in the PR

**Comment by dsyme** [comment]
I think the code looks good - the addition of FSharp.Compiler.UnitTests as well as the actual implementation of the feature.  

I'm not sure if this has made the Jan 16 deadline for F# 4.0 though?

**Comment by latkin** [comment]
This has been in-progress and earmarked for F# 4.0 for a long time. Given Marten's responsiveness to my requests and feedback over the past weeks/months (see original Codeplex thread) I think it's only fair to merge this for 4.0.

**Comment by dsyme** [comment]
Great, let's do it :)

**Comment by mrange** [comment]
FYI: Updated PR
1. Removed generated pp* files (@dsyme feedback)
2. Reverted lexhelp.* files (@mexx feedback)

**Comment by latkin** [comment]
Applied! Fantastic, Marten. Thanks for your work on this over the past 3 months. :clap:

if

**Comment by dsyme** [comment]
Fantastic!!!

**Comment by dsyme** [comment]
Hi @mrange,

We need to add a "Speclet" to this feature to [this collection](https://github.com/fsharp/FSharpLangDesign/tree/master/FSharp-4.0).  The format should follow the others: Links, Aim, Background, Design, Specification and so on. The "specification" section should give a guide for how [the F# Language Specification](http://fsharp.org/specs/language-spec/) will be updated for this feature.

Could you submit a PR to that repo with an initial design/spec document please, and we'll iterate with you.  If you need help doing the doc please let me know.

Thanks
Don

**Comment by mrange** [comment]
@dsyme - I threw together a speclet: https://github.com/fsharp/FSharpLangDesign/pull/40
